Automattic/wp-calypso
# composite-checkout: Remove useTotal from package API

In https://github.com/Automattic/wp-calypso/pull/81807 we removed a bunch of exports from the `@automattic/composite-checkout` package. Many were unused, some were leaky abstractions, and others didn't fit the vision of the package in its current form. One of the exports that was not removed was `useTotal` because it is still used in many places.

When `CheckoutProvider` was first implemented, the idea was that it would be provided a list of line items and a total in a standardized format and would then display those line items in a particular way automatically. However, as time passed it became clear that consumers needed full control over how line items were rendered, making the whole standardized line item feature a liability and a second source of truth. 

In https://github.com/Automattic/wp-calypso/pull/63720 we finally removed `useLineItems` from all payment methods. However, `useTotal` remained because most payment methods used it to display the total inside their submit button. That's been fixed by https://github.com/Automattic/wp-calypso/pull/82843. Now `useTotal` is completely unnecessary.

## Proposed Changes

This PR removes `useTotal` and the whole concept of standardized line items from the `@automattic/composite-checkout` package.

The only user-facing place still using these line items was the coupon, tax, and totals shown in checkout. This PR refactors those places to use the shopping cart data directly instead.

Because line items and totals are still needed by the `@automattic/composite-checkout` demo and tests, this PR refactors those places to have their own versions of line items.

## Testing Instructions

Not much testing should be required as this is already covered by tests and TypeScript. However, one place which we should verify is the user's displayed total in checkout, which was previously using the removed hook.

- Add a product (or even better, several products) to your cart and visit checkout.
- Verify that the coupon (if any), tax amount, and total displayed in the sidebar and near the submit button correctly displays the amount you owe. It should look identical to checkout before this PR.
- Leave checkout with the same items in the cart and click the masterbar cart (aka mini-cart) and verify its total is also correct and unchanged by this PR.

<img width="334" alt="Screenshot 2023-10-17 at 6 29 22 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/fdf796d6-0136-4023-846e-a0203d8e4fb4">

<img width="605" alt="Screenshot 2023-10-17 at 6 29 27 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/83172e0b-585e-4630-b623-9f8b9cc0ee44">

<img width="497" alt="Screenshot 2023-10-17 at 6 38 09 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/8819e0c1-7c12-40a9-9e67-1c77b5c0c8e1">
